src/java.base/share/classes/jdk/internal/foreign/ConfinedSession.java

```diff
@@ -101,7 +110,11 @@ void cleanup() {
             if (fst != ResourceCleanup.CLOSED_LIST) {
                 ResourceCleanup prev = fst;
                 fst = ResourceCleanup.CLOSED_LIST;
```

Comment:
Is there a reason why usage of fst here doesn't prevent successful escape analysis?

Reply:
So was the culprit `cleanup.next = fst;`?

Reply:
Yes, in combination with the while loop which runs all the cleanups, which adds a phi node that merges  the instance with `null`, at which point EA gives up.